File: kraken/src/main/java/glitch/kraken/services/request/FollowedVideosRequest.java
package glitch.kraken.services.request;

import glitch.api.http.HttpClient;
import glitch.api.http.HttpRequest;
import glitch.api.http.Routes;
import glitch.api.objects.enums.VideoType;
import glitch.auth.GlitchScope;
import glitch.auth.objects.json.Credential;
import glitch.kraken.object.enums.VideoSort;
import glitch.kraken.object.json.Video;
import glitch.kraken.object.json.collections.Videos;
import glitch.service.AbstractRequest;
import java.util.*;
import java.util.stream.Collectors;
import reactor.core.publisher.Mono;

public class FollowedVideosRequest extends AbstractRequest<Video, Videos> {
    private final Credential credential;
    private final Set<VideoType> videoType = new LinkedHashSet<>();
    private final Set<Locale> language = new LinkedHashSet<>();
    private Integer limit;
    private Integer offset;
    private VideoSort sort;

    public FollowedVideosRequest(HttpClient httpClient, Credential credential) {
        super(httpClient, Routes.get("/videos/followed").newRequest());
        this.credential = credential;
    }

    public FollowedVideosRequest videoType(VideoType... types) {
        return videoType(Arrays.asList(types));
    }

    public FollowedVideosRequest videoType(Collection<VideoType> types) {
        videoType.addAll(types);
        return this;
    }

    public FollowedVideosRequest language(String... languages) {
        return language(Arrays.stream(languages).map(Locale::forLanguageTag).collect(Collectors.toSet()));
    }

    public FollowedVideosRequest language(Locale... languages) {
        return language(Arrays.asList(languages));
    }

    public FollowedVideosRequest language(Collection<Locale> languages) {
        language.addAll(languages);
        return this;
    }

    public Mono<Videos> get() {
        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.USER_READ))
                .flatMap(b -> {
                    if (b) {

                        if (limit != null && limit > 0 && limit <= 100) {
                            request.queryParam("limit", limit);
                        }

                        if (offset != null && offset >= 0) {
                            request.queryParam("offset", offset);
                        }

                        if (!videoType.isEmpty()) {
                            request.queryParam("broadcast_type", videoType.stream().map(Enum::name).collect(Collectors.joining(",")).toLowerCase());
                        }

                        if (!language.isEmpty()) {
                            request.queryParam("language", language.stream().map(Locale::getLanguage).collect(Collectors.joining(",")).toLowerCase());
                        }

                        if (sort != null) {
                            request.queryParam("sort", sort.name().toLowerCase());
                        }

                        return httpClient.exchangeAs(request, Videos.class);
                    } else {
                        return Mono.error(handleScopeMissing(GlitchScope.USER_READ));
                    }
                });
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/request/FollowedVideosRequest.kt
package io.glitchlib.v5.service.request

import io.glitchlib.auth.Credential
import io.glitchlib.internal.http.AbstractRequest
import io.glitchlib.internal.http.HttpRequest
import io.glitchlib.model.VideoType
import io.glitchlib.v5.model.VideoSort
import java.util.*

// Scope.USER_READ
class FollowedVideosRequest internal constructor(private val credential: Credential) : AbstractRequest {
    //    AbstractRequest<Video, Videos>(httpClient, Routes.Companion.get("/videos/followed").newRequest()) {
    private val videoType = mutableSetOf<VideoType>()
    private val language = mutableSetOf<Locale>()
    private var limit: Int? = null
    private var offset: Int? = null
    private var sort: VideoSort? = null

    fun videoType(vararg types: VideoType) = videoType(types.toSet())

    fun videoType(types: Collection<VideoType>) = apply {
        videoType.addAll(types)
    }

    fun language(vararg languages: String) = language(languages.map(Locale::forLanguageTag))

    fun language(vararg languages: Locale) = language(Arrays.asList(*languages))

    fun language(languages: Collection<Locale>) = apply {
        language.addAll(languages)
    }

    fun limit(limit: Int) = apply {
        this.limit = limit
    }

    fun offset(offset: Int) = apply {
        this.offset = offset
    }

    fun sort(sort: VideoSort) = apply {
        this.sort = sort
    }

    override fun invoke(): HttpRequest.() -> Unit = {
        if (limit != null && limit!! > 0 && limit!! <= 100) {
            addQueryParameters("limit", limit!!.toString())
        }

        if (offset != null && offset!! >= 0) {
            addQueryParameters("offset", offset!!.toString())
        }

        if (!videoType.isEmpty()) {
            addQueryParameters("broadcast_type", videoType.joinToString(",") { it.name }.toLowerCase())
        }

        if (!language.isEmpty()) {
            addQueryParameters("language", language.joinToString(",") { it.language }.toLowerCase())
        }

        if (sort != null) {
            addQueryParameters("sort", sort!!.name.toLowerCase())
        }
    }
}
